#!/usr/bin/env Rscript

# Copyright by Daniel Loos
#
# Research Group Systems Biology and Bioinformatics - Head: Assoc. Prof. Dr. Gianni Panagiotou
# https://www.leibniz-hki.de/en/systembiologie-und-bioinformatik.html
# Leibniz Institute for Natural Product Research and Infection Biology - Hans Knöll Institute (HKI)
# Adolf-Reichwein-Straße 23, 07745 Jena, Germany
#
# The project code is licensed under BSD 2-Clause.
# See the LICENSE file provided with the code for the full license.

#
# create NCBI project to integrate in DAnIEL
# - requires sra sub dir and meta.csv created by grabseqs sra
#

library(tidyverse)
library(magrittr)
library(jsonlite)
library(rentrez)
library(xml2)
library(optparse)

args <- base::list(
  optparse::make_option(
    opt_str = "--project-dir",
    type = "character",
    help = "Path to project directory containing sub dir sra and meta.csv generated by grabseqs sra"
  ),
  optparse::make_option(
    opt_str = "--project-template-json",
    type = "character",
    help = "Path to DAnIEL template project.json"
  ),
  optparse::make_option(
    opt_str = "--min-reads",
    type = "integer",
    default = 1000,
    help = "Minimum number of spots per fastq file"
  )
) %>%
  optparse::OptionParser(
    option_list = .,
    description = "Make projects downloaded by grabseqs sra acessible in DAnIEL"
  ) %>%
  optparse::parse_args(convert_hyphens_to_underscores = TRUE)

# args <- list(
#   project_dir = "/sbidata/dloos/prj/6-its-web-server/db/projects/PRJNA286273/",
#   project_template_json = "~/prj/6-its-web-server/src/etc/default_project.json",
#   min_reads = 1000
# )

sra_meta_tbl <-
  args$project_dir %>%
  paste0("/meta.csv") %>%
  read_csv() %>%
  filter(spots >= args$min_reads)

sra_ids <- sra_meta_tbl$Run

# build json config
project_l <- read_json(args$project_template_json)
project_l$input$sra_ids <- sra_ids
project_l$input$local_samples <- sra_ids
project_l$project_dir <- args$project_dir

# write project.json
input_dir <- paste0(args$project_dir, "/input/")
reads_dir <- paste0(input_dir, "reads/")

dir.create(reads_dir, recursive = TRUE)
write_json(project_l, paste0(input_dir, "project.json"), pretty = TRUE)

# create samples.csv
rentrez::set_entrez_key("720771ca7b2856890294bba45c829f9e8a08")

get_biosample_attributes <- function(biosample_id) {
  Sys.sleep(0.1) # do not spam NCBI

  biosample_id %>%
    str_remove_all("A-z") %>%
    entrez_fetch(db = "biosample", id = ., rettype = "xml") %>%
    read_xml() %>%
    as_list() %>%
    pluck("BioSampleSet", "BioSample", "Attributes") %>%
    lapply(function(x) {
      tibble(
        biosample_id = biosample_id,
        key = x %>% attributes() %>% pluck("display_name") %>% as.character(),
        val = x %>% as.character()
      )
    }) %>%
    bind_rows() %>%
    return()
}

biosamples_tbl <-
  sra_meta_tbl %>%
  select(sample_id = Run, biosample_id = BioSample) %>%
  {
    map(.$biosample_id, get_biosample_attributes)
  } %>%
  bind_rows() %>%
  spread(key, val)

samples_tbl <-
  sra_meta_tbl %>%
  select(sample_id = Run, biosample_id = BioSample) %>%
  left_join(biosamples_tbl, by = "biosample_id")

write_csv(samples_tbl, paste0(input_dir, "samples.csv"))

# link read files
link_read_file <- function(sra_id, mate = "_1") {
  file.symlink(paste0(args$project_dir, "sra/", sra_id, mate, ".fastq.gz"), paste0(reads_dir, sra_id, mate, ".raw.fq.gz"))
}
map(sra_ids, ~ link_read_file(.x, "_1")) %>% invisible()
map(sra_ids, ~ link_read_file(.x, "_2")) %>% invisible()
